versiondb/client/changeset.go

```diff
@@ -0,0 +1,363 @@
+package client
+
+import (
+	"bufio"
+	"compress/zlib"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"math/bits"
+	"os"
+	"sort"
+	"strings"
+
+	"github.com/cosmos/iavl"
+	"github.com/golang/snappy"
+)
+
+const (
+	ZlibFileSuffix   = ".zz"
+	SnappyFileSuffix = ".snappy"
+)
+
+// WriteChangeSet writes a version of change sets to writer.
+//
+// Change set file format:
+// ```
+// version: int64
+// size: int64         // size of whole payload
+// payload:
+//
+//	delete: int8
+//	keyLen: varint-uint64
+//	key
+//	[ // if delete is false
+//	  valueLen: varint-uint64
+//	  value
+//	]
+//	repeat with next key-value pair
+//
+// repeat with next version
+// ```
+func WriteChangeSet(writer io.Writer, version int64, cs iavl.ChangeSet) error {
+	var size int
+	items := make([][]byte, 0, len(cs.Pairs))
+	for _, pair := range cs.Pairs {
+		buf, err := encodeKVPair(pair)
+		if err != nil {
+			return err
+		}
+		size += len(buf)
+		items = append(items, buf)
+	}
+
+	var versionHeader [16]byte
+	binary.LittleEndian.PutUint64(versionHeader[:], uint64(version))
+	binary.LittleEndian.PutUint64(versionHeader[8:], uint64(size))
+
+	if _, err := writer.Write(versionHeader[:]); err != nil {
+		return err
+	}
+	for _, item := range items {
+		if _, err := writer.Write(item); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ReadChangeSet decode a version of change set from reader.
+// if parseChangeset is false, it'll skip change set payload directly.
+//
+// returns (version, number of bytes read, changeSet, err)
+func ReadChangeSet(reader Reader, parseChangeset bool) (int64, int64, *iavl.ChangeSet, error) {
+	var versionHeader [16]byte
+	if _, err := io.ReadFull(reader, versionHeader[:]); err != nil {
+		return 0, 0, nil, err
+	}
+	version := binary.LittleEndian.Uint64(versionHeader[:8])
+	size := int64(binary.LittleEndian.Uint64(versionHeader[8:16]))
+	var changeSet iavl.ChangeSet
+
+	if size == 0 {
+		return int64(version), 16, &changeSet, nil
```

Comment:
the version is unlikely to reach 2^63-1 soon, right?

Reply:
it takes hundreds of years to even overflow uint32 ;D